src/mame/sinclair/specnext_sprites.cpp

```diff
@@ -0,0 +1,282 @@
+            spr_cur.paloff = BIT(spr_cur_attr[2], 4, 4);
+            spr_cur.pattern = BIT(spr_cur_attr[3], 0, 6);
+
+            spr_cur.h = BIT(spr_cur_attr[4], 7) && BIT(sprite_attr[3], 6);
+            bool spr_cur_n6 = BIT(spr_cur_attr[4], 6) && spr_cur.h;
+
+            spr_cur.yscale = BIT(spr_cur_attr[4], 1, 2);
+            spr_cur.xscale = BIT(spr_cur_attr[4], 3, 2);
+            spr_cur.rel_type = BIT(spr_cur_attr[4], 5);
+            if (spr_cur.h)
+                spr_cur.pattern = (spr_cur.pattern << 1) | spr_cur_n6;
+
+            m_sprites_cache.push_back(spr_cur);
+        }
+    }
+}
+
+void specnext_sprites_device::update_config()
+{
+    if (gfx(0) == nullptr) return;
+
+    // Change granularity to 16 for 8bpp
+    gfx(0)->set_granularity(16);
+    gfx(2)->set_granularity(16);
+
+    m_clip_window = m_over_border
+        ? (m_border_clip_en == 0 )
+            ? SCREEN_AREA
+            : rectangle { m_clip_x1 << 1, (m_clip_x2 << 1) | 1, m_clip_y1, m_clip_y2 }
+        : rectangle { m_clip_x1 + OVER_BORDER, m_clip_x2 + OVER_BORDER, m_clip_y1 + OVER_BORDER, m_clip_y2 + OVER_BORDER };
+    m_clip_window.set_origin(m_offset.first, m_offset.second);
+}
+
+void specnext_sprites_device::io_w(offs_t addr, u8 data)
+{
+    bool attr_num_change = 0;
+    if (addr == 0x5b)
+    {
+        // Ugh
+        gfx(0)->mark_all_dirty();
+        gfx(1)->mark_all_dirty();
+        gfx(2)->mark_all_dirty();
+        gfx(3)->mark_all_dirty();
+
+        m_sprite_pattern_ram[m_pattern_index] = data;
+        m_pattern_index++;
+    }
+    else if (addr == 0x303b)
+    {
+        m_pattern_index = (BIT(data, 0, TOTAL_PATTERN_BITS) << 8) | (data & 0x80);
+        m_attr_index = BIT(data, 0, TOTAL_SPRITES_BITS) << 3;
+        attr_num_change = 1;
+    }
+    else if (addr == 0x57)
+    {
+        m_sprites_cache.clear();
+        m_sprite_attr_ram[m_attr_index] = data;
+
+        const bool index_inc_attr_by_8 = BIT(m_attr_index, 2) || ((BIT(m_attr_index, 0, 3) == 0b011) && (BIT(data, 6) == 0));
+        const u16 index_inc_in_s = index_inc_attr_by_8
+            ? BIT(m_attr_index, 3, TOTAL_SPRITES_BITS)
+            : m_attr_index;
+        const u16 index_inc_out_s = index_inc_in_s + 1;
+
+        if (index_inc_attr_by_8 == 0)
+            m_attr_index = BIT(index_inc_out_s, 0, TOTAL_SPRITES_BITS + 3);
+        else
+        {
+            m_attr_index = BIT(index_inc_out_s, 0, TOTAL_SPRITES_BITS) << 3;
+            attr_num_change = 1;
+        }
+    }
+
+    if (attr_num_change && m_mirror_tie)
+    {
+        m_mirror_sprite_q &= ~0xff;
+        m_mirror_sprite_q |= BIT(m_attr_index, 3, TOTAL_SPRITES_BITS);
+        m_mirror_sprite_q |= m_pattern_index & 0x80;
+    }
+}
+
+void specnext_sprites_device::mirror_data_w(u8 mirror_data)
+{
+    bool mirror_num_change = 0;
+    if (m_mirror_index == 0b111)
+    {
+        m_mirror_sprite_q = mirror_data;
+        mirror_num_change = 1;
+    }
+    else if (m_mirror_inc) // ...
```

Comment:
What’s the `...` comment supposed to mean?  This doesn’t tell the person reading the code anything, but suggests there’s something they’re supposed to know.